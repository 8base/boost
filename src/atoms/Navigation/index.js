export { Navigation, theme } from './Navigation';