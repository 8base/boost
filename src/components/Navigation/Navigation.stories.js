import React from 'react';

export default (asStory) => {
  asStory('Components/Navigation', module, (story, { Navigation }) => {
    story
      .add('without groups', () => (
        <Navigation color="GREEN">
          <Navigation.Item icon="Trashcan" label="Jobs" to="/jobs" />
          <Navigation.Item icon="Mail" label="Posted" to="/posted" />
          <Navigation.Item icon="Search" label="My Active Jobs" to="/my-active-jobs" />
        </Navigation>
      ));
  });
};

