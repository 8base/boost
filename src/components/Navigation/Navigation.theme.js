// @flow
import fp from 'lodash/fp';
import color from 'color';

import { createThemeTag } from '../../theme/createThemeTag';

const name = 'navigation';

const [NavigationTag, themeNavigation] = createThemeTag(name, ({ COLORS }) => ({
  root: ({ expanded, collapsedWidth, expandedWidth }) => ({
    display: 'flex',
    flexDirection: 'column',
    height: '100%',
    transition: 'width 0.3s',

    ...(expanded ? {
      '.NavigationItem-label': {
        transitionDelay: '0.15s',
        opacity: '1',
        visibility: 'visible',
      },
      width: expandedWidth,
    } : {
      width: collapsedWidth,
    }),
  }),
  modifiers: {
    color: fp.mapValues((color) => ({
      backgroundColor: color,
    }), COLORS),
  },
}));


const [NavigationItemTag, themeItemMain] = createThemeTag(`${name}Item`, ({ COLORS }) => ({
  root: props => ({
    cursor: 'pointer',
    position: 'relative',
    textDecoration: 'none',
    userSelect: 'none',
    display: 'flex',
    color: COLORS.LIGHT_PRIMARY_TEXT_COLOR,

    '&:hover, &.active': {
      backgroundColor: color(COLORS[props.color]).darken(0.2).hex(),
    },
  }),
  modifiers: {
    color: fp.mapValues((color) => ({
      backgroundColor: color,
    }), COLORS),
  },
}));

const [NavigationItemIcon, themeItemIcon] = createThemeTag(`${name}ItemIcon`, ({ COLORS }) => ({
  root: {
    display: 'flex',
    width: '56px',
    height: '56px',
    flexShrink: '0',
    justifyContent: 'center',
    alignItems: 'center',
    color: COLORS.LIGHT_PRIMARY_TEXT_COLOR,
  },
}));

const [NavigationItemLabel, themeItemLabel] = createThemeTag(`${name}ItemLabel`, {
  root: {
    display: 'flex',
    alignItems: 'center',
    textTransform: 'uppercase',
    fontWeight: 600,
    paddingRight: '20px',
    opacity: '0',
    visibility: 'hidden',
    transition: 'all 0.1s',
    whiteSpace: 'nowrap',
  },
});


const [NavigationItemLabelPreview, themeItemLabelPreview] = createThemeTag(`${name}ItemLabelPreview`, {
  root: {
    alignItems: 'center',
    textTransform: 'uppercase',
    fontWeight: 600,
  },
});


const theme = {
  ...themeNavigation,
  ...themeItemMain,
  ...themeItemLabel,
  ...themeItemIcon,
  ...themeItemLabelPreview,
};

export {
  theme,
  NavigationTag,
  NavigationItemTag,
  NavigationItemLabel,
  NavigationItemIcon,
  NavigationItemLabelPreview,
};
