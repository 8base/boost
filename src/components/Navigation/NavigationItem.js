import React from 'react';
import color from 'color';
import fp from 'lodash/fp';

import { Icon } from '../Icon';
import { createStyledTag, createComponentTheme } from '../../utils';
import { PALETTE } from '../../theme';

type NavigationItemProps = {|
  exact?: boolean,
  to?: string,
  component?: React$Node,
  color?: $Keys<typeof PALETTE>
|};

const name = 'navigationItem';

const themeMain = createComponentTheme(name, ({ COLORS }) => ({
  root: {
    color: COLORS.LIGHT_PRIMARY_TEXT_COLOR,
  },
  modifiers: {
    color: fp.mapValues((color) => ({
      backgroundColor: color,
    }), PALETTE),
  },
}));

const themeLabel = createComponentTheme(`${name}Label`, ({ COLORS }) => ({
  color: COLORS.LIGHT_PRIMARY_TEXT_COLOR,
}));

const theme = {
  ...themeMain,
  ...themeLabel,
};

const StyledTag = createStyledTag(name, (props) => ({
  cursor: 'pointer',
  position: 'relative',
  textDecoration: 'none',
  userSelect: 'none',
  display: 'flex',

  '&:hover, &.active': {
    backgroundColor: color(PALETTE[props.color]).darken(0.2).hex(),
  },
}));

const NavigationItemIcon = createStyledTag(`${name}Icon`, {
  display: 'flex',
  width: '60px',
  height: '60px',
  justifyContent: 'center',
  alignItems: 'center',
});

const NavigationItemLabel = createStyledTag(`${name}Label`, {
  display: 'none',
  alignItems: 'center',
  textTransform: 'uppercase',
  fontWeight: 600,
  paddingRight: '20px',
});

const NavigationItem = ({ icon, label, ...rest }: NavigationItemProps) => (
  <StyledTag { ...rest }>
    <NavigationItemIcon>
      <Icon name={ icon } color="WHITE" />
    </NavigationItemIcon>
    { label && <NavigationItemLabel className="NavigationItem-label">{ label }</NavigationItemLabel> }
  </StyledTag>
);

NavigationItem.defaultProps = {
  ...theme[name].defaults,
  tagName: 'a',
};

export { NavigationItem, theme };
