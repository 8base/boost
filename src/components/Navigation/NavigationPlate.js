import React from 'react';
import fp from 'lodash/fp';

import { createStyledTag, createComponentTheme } from '../../utils';
import { PALETTE } from '../../theme';

type NavigationPlateProps = {|
  children?: React.Node,
  color: $Keys<typeof PALETTE>
|};

const name = 'navigationPlate';

const theme = createComponentTheme(name, {
  root: {
    width: '60px',
  },
  modifiers: {
    color: fp.mapValues((color) => ({
      backgroundColor: color,
    }), PALETTE),
  },
});

const StyledTag = createStyledTag(name, {
  display: 'flex',
  flexDirection: 'column',
  height: '100%',

  '&:hover': {
    width: 'auto',

    '.NavigationItem-label': {
      display: 'flex',
    },
  },
});

function NavigationPlate({
  children,
  color,
  ...rest
  }: NavigationPlateProps) {
  return (
    <StyledTag { ...rest } color={ color } tagName="nav">
      {
        React.Children.map(children, child =>
          React.cloneElement(child, { color: child.props.color || color }),
        )
      }
    </StyledTag>
  );
}

export { NavigationPlate, theme };
