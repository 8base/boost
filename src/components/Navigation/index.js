// @flow

export { Navigation } from './Navigation';
export { theme } from './Navigation.theme';
