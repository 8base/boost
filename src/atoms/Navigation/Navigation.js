import { NavigationGroup as Group, theme as navigationGroupTheme } from './NavigationGroup';
import { NavigationItem as Item, theme as navigationItemTheme } from './NavigationItem';
import { NavigationPlate as Plate, theme as navigationPlateTheme } from './NavigationPlate';

const Navigation = {
  Group,
  Item,
  Plate,
};

const theme = {
  ...navigationGroupTheme,
  ...navigationItemTheme,
  ...navigationPlateTheme,
};

export {
  Navigation,
  theme,
};
