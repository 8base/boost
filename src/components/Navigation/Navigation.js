// @flow

import React from 'react';
import { COLORS } from '../../theme';

import { NavigationTag } from './Navigation.theme';
import { NavigationItem } from './NavigationItem';


type NavigationProps = {
  children?: React$Node,
  expandedWidth?: string,
  collapsedWidth: string,
  color: $Keys<typeof COLORS>,
  expanded?: boolean,
};

type NavigationState = {
  expandedWidth?: string,
  expanded: boolean,
};

class Navigation extends React.Component<NavigationProps, NavigationState> {
  static displayName = 'Navigation';
  static Item = NavigationItem;
  static defaultProps = {
    collapsedWidth: '60px',
  };

  state = {
    expandedWidth: this.props.expandedWidth,
    expanded: false,
  }

  // $FlowFixMe
  navRef = React.createRef();

  componentDidMount() {
    const { current: navElement } = this.navRef;

    if (navElement && !this.isControllable()) {
      requestAnimationFrame(() => {
        if (!this.props.expandedWidth) {
          this.setState({
            expandedWidth: getComputedStyle(navElement).width,
          });
        }

        navElement.style.width = this.props.collapsedWidth;
        navElement.addEventListener('mouseenter', this.onMouseEnter);
        navElement.addEventListener('mouseleave', this.onMouseLeave);
      });
    }
  }

  componentWillUnmount() {
    const { current: navElement } = this.navRef;

    if (navElement && !this.isControllable()) {
      navElement.removeEventListener('mouseenter', this.onMouseEnter);
      navElement.removeEventListener('mouseleave', this.onMouseLeave);
    }
  }

  isControllable = () => typeof this.props.expanded === 'boolean';

  onMouseEnter = () => {
    if (this.navRef.current) {
      this.navRef.current.style.width = this.state.expandedWidth;
    }

    this.setState({
      expanded: true,
    });
  }

  onMouseLeave = () => {
    if (this.navRef.current) {
      this.navRef.current.style.width = this.props.collapsedWidth;
    }

    this.setState({
      expanded: false,
    });
  }

  render() {
    const { color, children, ...rest } = this.props;

    const controllable = this.isControllable();
    const expanded = controllable ? this.props.expanded : this.state.expanded;

    return (
      <NavigationTag { ...rest } expanded={ expanded } controllable={ controllable } insideRef={ this.navRef } color={ color } tagName="nav">
        {
          React.Children.map(children, child =>
            React.cloneElement(child, {
              color: child.props.color || color,
              parentExpanded: expanded,
            }),
          )
        }
      </NavigationTag>
    );
  }
}

export { Navigation };
