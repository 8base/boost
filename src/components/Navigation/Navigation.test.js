// @flow
import React from 'react';
import { Navigation } from './Navigation';

describe('<Navigation />', () => {
  it('should render navigation with correct props', () => {
    const wrapper = mount(
      <Navigation.Plate color="GREEN">
        <Navigation.Item icon="Trashcan" label="Jobs" href="/jobs" />
      </Navigation.Plate>,
    );

    expect(wrapper.find('a').props().href).toBe('/jobs');
  });

  it('should render navigation with custom tag', () => {
    const Link = () => null;

    const wrapper = mount(
      <Navigation.Plate color="GREEN">
        <Navigation.Item icon="Trashcan" label="Jobs" tagName={ Link } to="/jobs" />
      </Navigation.Plate>,
    );

    expect(wrapper.find(Link).props().to).toBe('/jobs');
  });
});
