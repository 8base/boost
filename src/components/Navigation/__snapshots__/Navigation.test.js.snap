// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> should shallow navigation 1`] = `
<Boost(navigation)
  collapsedWidth="60px"
  color="GREEN"
  insideRef={
    Object {
      "current": null,
    }
  }
  tagName="nav"
>
  <NavigationItem
    color="GREEN"
    href="/jobs"
    icon="Trashcan"
    key=".0"
    label="Jobs"
    parentExpanded={false}
    tagName="a"
  />
</Boost(navigation)>
`;

exports[`<Navigation /> should shallow navigation item 1`] = `
<Boost(navigationItem)
  href="/jobs"
  tagName="a"
>
  <Boost(navigationItemIcon)
    modifiers={
      Object {
        "href": "/jobs",
        "tagName": "a",
      }
    }
    tagName="div"
  >
    <Icon
      color="WHITE"
      name="Trashcan"
      size="md"
    />
  </Boost(navigationItemIcon)>
  <Boost(navigationItemLabel)
    className="NavigationItem-label"
    modifiers={
      Object {
        "href": "/jobs",
        "tagName": "a",
      }
    }
    tagName="div"
  >
    Jobs
  </Boost(navigationItemLabel)>
</Boost(navigationItem)>
`;
